### DISCUSSION TITLE
Webpack Plugin's Module Size

### DISCUSSION THREAD

**Discussion by Pirororor** [webpack/webpack#19054] [Q&A] [answered]
I am currently trying to monitor a specific module's output bundle size that is in an asset.

I tried creating a webpack plugin to simply log down all the modules' sizes:

```
apply(compiler) {
  compiler.hooks.done.tapAsync('ModuleSizePlugin', (stats, callback) => {
        const moduleBreakdown = {};
        const compilation = stats.compilation.chunks.forEach(chunk => {
          const chunkInfo = {
            id: chunk.id,
            name: chunk.name,
            modules: [],
          };
  
          chunk.getModules().forEach(module => {
            const moduleInfo = {
              resource: module.resource || '',
              name: module.name || '',
              module: this.getModuleSize(module),
              chunk: {
                id: chunk.id,
                name: chunk.name,
              },
            };
            chunkInfo.modules.push(moduleInfo);
            if (!moduleBreakdown[module.resource]) {
              moduleBreakdown[module.resource] = { info: [] };
            }
            moduleBreakdown[module.resource].info.push(moduleInfo);
          });
        });    
  
       console.log(moduleInfo)
  }
}


getModuleSize(module) {
  const ret = {};
  if (module._source && module._source.size) {
    ret['_sourceSize'] = module._source.size();
  }
  if (module.size) {
    ret['size'] = module.size();
  }
  return ret;
}
```

1. I that there  are `_source.size`, `size` and a `sizes.javascript` fields, but I realise that not all modules have all 3. When does each appear?
2. It seems that `size` does not correspond to the size in the asset. (e.g when the asset only contains only 1 module, seems that the module's size is larger than the asset's size.) When is the size field in the module calculated?

**Comment by alexander-akait** [comment]
> It seems that size does not correspond to the size in the asset. (e.g when the asset only contains only 1 module, seems that the module's size is larger than the asset's size.) When is the size field in the module calculated?

For `assets` there is `sizes.asset` 

Code - https://github.com/webpack/webpack/blob/main/lib/asset/AssetGenerator.js#L591

Same for any modules

**Reply by Pirororor** [reply]
> > It seems that size does not correspond to the size in the asset. (e.g when the asset only contains only 1 module, seems that the module's size is larger than the asset's size.) When is the size field in the module calculated?
> 
> For `assets` there is `sizes.asset`
> 
> Code - https://github.com/webpack/webpack/blob/main/lib/asset/AssetGenerator.js#L591
> 
> Same for any modules

I see. However, if the output bundle asset consists of more than 1 module, is there a way to see the breakdown on how much each bundle occupied?

**Reply by alexander-akait** [reply]
The module has https://github.com/webpack/webpack/blob/main/lib/Module.js#L876C2-L876C16

**Reply by Pirororor** [reply, marked as answer]
Thanks a lot for your help.

> It seems that size does not correspond to the size in the asset. (e.g when the asset only contains only 1 module, seems that the module's size is larger than the asset's size.) When is the size field in the module calculated?

This is the issue I face when using module's `size` field. 

After researching more, seems that `module`'s size field is pre-minification and pre-bundling, and the size directly corresponds to the source's size. Since I am lookign for how much each module's size takes out in the **output bundle**, seems like webpack-bundle-analyzer's way of manually estimating how much size each module uses up in a bundle is the way to go.
